src/NuGet.CommandLine/MsBuildUtility.cs

```diff
@@ -115,86 +115,178 @@ public static IEnumerable<string> GetAllProjectFileNames(string solutionFile, st
+                        var lines = output.Split(
+                            new[] { Environment.NewLine },
+                            StringSplitOptions.RemoveEmptyEntries);
 
-                        if (process.ExitCode == 0)
+                        string versionString = lines.LastOrDefault(
+                            line => !string.IsNullOrWhiteSpace(line));
+                        Version version;
+                        if (Version.TryParse(versionString, out version))
                         {
-                            var output = process.StandardOutput.ReadToEnd();
-
-                            // The output of msbuid /version with MSBuild 14 is:
-                            //
-                            // Microsoft (R) Build Engine version 14.0.23107.0
-                            // Copyright(C) Microsoft Corporation. All rights reserved.
-                            //
-                            // 14.0.23107.0                            
-                            var lines = output.Split(
-                                new[] { Environment.NewLine },
-                                StringSplitOptions.RemoveEmptyEntries);
-
-                            version = lines.LastOrDefault(
-                                line => !string.IsNullOrWhiteSpace(line));
+                            return version;
                         }
                     }
                 }
-                catch
-                {
-                    // ignore errors
-                }
+            }
+            catch
+            {
+                // ignore errors                
             }
 
-            var versionString = version.Contains('.') ?
-                version :
-                version + ".0";
-            Version ver;
-            if (!Version.TryParse(versionString, out ver))
+            return null;
+        }
+
+        /// <summary>
+        /// Gets the msbuild toolset that matches the given <paramref name="msbuildVersion"/>.
+        /// </summary>
+        /// <param name="msbuildVersion">The msbuild version. Can be null.</param>
+        /// <param name="installedToolsets">List of installed toolsets, 
+        /// ordered by ToolsVersion, from highest to lowest.</param>
+        /// <returns>The matching toolset.</returns>
+        public static Toolset SelectMsbuildToolset(
+            Version msbuildVersion,
+            IList<Toolset> installedToolsets)
+        {
+            Toolset selectedToolset;
+            if (msbuildVersion == null)
             {
-                var message = string.Format(
-                    CultureInfo.CurrentCulture,
-                    LocalizedResourceManager.GetString(
-                        nameof(NuGetResources.Error_InvalidMsbuildVersion)),
-                    version);
+                // MSBuild does not exist in PATH. In this case, the highest installed version is used
+                selectedToolset = installedToolsets.First();
```

Comment:
i think the best approach here is to use FirstOrDefault and control the error we are showing the user